pageserver/src/tenant.rs

```diff
@@ -3643,28 +3642,135 @@ impl Tenant {
             .schedule_index_upload_for_full_metadata_update(&metadata)
             .context("branch initial metadata upload")?;
 
+        // At this point we have dropped our guard on [`Self::timelines_creating`], and
+        // the timeline is visible in [`Self::timelines`], but it is _not_ durable yet.  We must
+        // not send a success to the caller until it is.  The same applies to handling retries,
+        // that is done in [`Self::start_creating_timeline`].
+        new_timeline
+            .remote_client
+            .wait_completion()
+            .await
+            .context("wait for timeline initial uploads to complete")?;
+
+        match activate {
+            ActivateTimelineArgs::Yes { broker_client } => {
+                new_timeline.activate(self.clone(), broker_client, None, ctx);
+            }
+            #[cfg(test)]
+            ActivateTimelineArgs::No => {}
+        }
```

Comment:
Is there a reason for duplicating this creation epilogue? It could be done once in `Tenant::create_timeline`

Reply:
Yes there is. The `Tenant::create_timeline` can't tell whether the returned `Arc<Timeline>` was newly created (this code path) or whether we returned an `Arc::clone` of the already-existing Timeline which we deemed idempotent earlier.

Reply:
Okay - I see it now. If you wanted to, you could wrap `Arc<Timeline> into an enum with two variants `Created` and `PreExisting`, but I don't mind, so up to you.

Reply:
commit 6fc9744a6001a9a015edf0ce55daffb01f8d8fb2 (HEAD -> problame/refactor-timeline-create-args)
Author: Christian Schwarz <christian@neon.tech>
Date:   Thu Oct 24 14:38:58 2024 +0200

    DRY the wait_completion + activation (+ reverse their order)